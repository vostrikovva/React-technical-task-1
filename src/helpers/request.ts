import { API_URL } from '../config'

const request = async (url: string, options?: RequestInit | undefined) => {
  const result = await fetch(API_URL + url, options)
  const data = await result.json()

  return data
}

export default request
