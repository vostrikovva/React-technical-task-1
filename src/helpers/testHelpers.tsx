import { render } from '@testing-library/react'
import { Provider } from 'react-redux'

import { store } from '../store/store'

export const renderWithStore = (component: React.ReactElement) => render(<Provider store={store} >{component}</Provider>)
