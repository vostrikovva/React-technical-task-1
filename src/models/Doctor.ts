import Address from './Address'

export default interface Doctor {
  id: string
  name: string
  address?: Address
}

export interface BookDoctorBody {
  doctorId: string
  date: string
}

export interface BookDoctorResponse {
  doctorId: string
  date: string
  createdAt: string
  updatedAt: string
  id: string
}
