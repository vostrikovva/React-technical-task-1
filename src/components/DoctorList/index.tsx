import { FC, memo, useState } from 'react'

import Doctor from '../../models/Doctor'
import DoctorBooking from '../DoctorBooking'
import DoctorInfo from '../DoctorInfo'

interface DoctorInfoProps {
  doctors: Doctor[]
}

const DoctorList: FC<DoctorInfoProps> = ({ doctors }) => {
  const [bookedDoctor, setBookedDoctor] = useState<Doctor | null>(null)

  return (
    <>
      {bookedDoctor !== null && <DoctorBooking doctor={bookedDoctor} setBookedDoctor={setBookedDoctor} />}
      {doctors.map((doctor) => <DoctorInfo key={doctor.id} doctor={doctor} setBookedDoctor={setBookedDoctor} />)}
    </>
  )
}

export default memo(DoctorList)
