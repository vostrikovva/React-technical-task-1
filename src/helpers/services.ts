import API_ROUTES from '../constants/APIRoutes'
import Availability from '../models/Availability'
import Doctor, { BookDoctorBody, BookDoctorResponse } from '../models/Doctor'
import request from './request'

const services = {
  fetchDoctors: async () => await request(API_ROUTES.doctors) as Doctor[],
  fetchAvailabilities: async (doctorId: string) => await request(`${API_ROUTES.availabilities}?doctorId=${doctorId}`) as Availability[],
  bookDoctor: async (body: BookDoctorBody) => await request(API_ROUTES.bookDoctor, {
    method: 'POST',
    body: JSON.stringify(body),
    headers: {
      'Content-Type': 'application/json'
    }
  }) as BookDoctorResponse
}

export default services
