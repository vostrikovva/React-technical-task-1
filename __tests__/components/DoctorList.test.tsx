import { render, cleanup } from '@testing-library/react'
import rerender from 'react-test-renderer'
import userEvent from '@testing-library/user-event'

import '@testing-library/jest-dom'

import Doctor from '../../src/models/Doctor'
import DoctorList from '../../src/components/DoctorList'
import { renderWithStore } from '../../src/helpers/testHelpers'

const doctors: Doctor[] = [{
  id: '957dc48d-4691-47e7-b1fe-82c91b99d3b2',
  name: 'Dr. Hortense Benoit',
  address: {
    line1: '551 Élisée de la Bûcherie',
    line2: 'Suite 975',
    postalCode: '75830',
    city: 'Le gallhaven',
    country: 'GU'
  }
},
{
  id: 'b76aadc6-bc92-4400-a75a-8c3506353779',
  name: 'Dr. Abdonie Adam',
  address: {
    line1: '050 Garnier de la Paix',
    line2: 'Suite 124',
    postalCode: '44493',
    city: 'Lake Pénélope',
    country: 'PA'
  }
}]

jest.mock('../../src/helpers/services', () => ({
  fetchAvailabilities: () => [{
    start: '2023-01-05T16:35:52.918Z'
  },
  {
    start: '2023-01-21T06:41:58.114Z'
  },
  {
    start: '2023-01-15T09:14:07.535Z'
  }],
  bookDoctor: () => ({ createdAt: '2023-01-15T09:14:07.535Z' })
}))

describe('DoctorList component', () => {
  afterEach(() => {
    cleanup()
  })

  test('Render the component', async () => {
    const component = render(<DoctorList doctors={doctors} />)

    expect(await component.findByText(doctors[0].name)).toBeInTheDocument()
  })

  test('Snapshot test', () => {
    const tree = rerender.create(<DoctorList doctors={doctors} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Test behaviour after booking clicking', async () => {
    const component = renderWithStore(<DoctorList doctors={doctors} />)

    const firstRender = component.asFragment()
    const button = component.getAllByText('Booking')
    await userEvent.click(button[1])

    expect(await component.findByText('Booking the doctor Dr. Abdonie Adam')).toBeInTheDocument()
    expect(firstRender).not.toBe(component.asFragment())
  })
})
