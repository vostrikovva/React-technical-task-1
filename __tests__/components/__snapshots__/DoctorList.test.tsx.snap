// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoctorList component Snapshot test 1`] = `
[
  <div
    className="root m-2 card"
  >
    <div
      className="card-body"
    >
      <div
        className="mb-3 card-title h5"
      >
        Dr. Hortense Benoit
      </div>
      <div
        className="row mb-10"
      >
        <div
          className="col-auto d-flex align-items-center"
        >
          <div
            className="card-subtitle h6"
          >
            Address
          </div>
        </div>
        <div
          className="col"
        >
          <p
            className="mb-0 card-text"
          >
            GU, Le gallhaven
          </p>
          <p
            className="mb-0 card-text"
          >
            Postal Code: 
            75830
          </p>
          <p
            className="mb-0 card-text"
          >
            Address line 1: 
            551 Élisée de la Bûcherie
          </p>
          <p
            className="mb-0 card-text"
          >
            Address line 2: 
            Suite 975
          </p>
        </div>
      </div>
      <div
        className="mt-2 d-flex justify-content-end"
      >
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Booking
        </button>
      </div>
    </div>
  </div>,
  <div
    className="root m-2 card"
  >
    <div
      className="card-body"
    >
      <div
        className="mb-3 card-title h5"
      >
        Dr. Abdonie Adam
      </div>
      <div
        className="row mb-10"
      >
        <div
          className="col-auto d-flex align-items-center"
        >
          <div
            className="card-subtitle h6"
          >
            Address
          </div>
        </div>
        <div
          className="col"
        >
          <p
            className="mb-0 card-text"
          >
            PA, Lake Pénélope
          </p>
          <p
            className="mb-0 card-text"
          >
            Postal Code: 
            44493
          </p>
          <p
            className="mb-0 card-text"
          >
            Address line 1: 
            050 Garnier de la Paix
          </p>
          <p
            className="mb-0 card-text"
          >
            Address line 2: 
            Suite 124
          </p>
        </div>
      </div>
      <div
        className="mt-2 d-flex justify-content-end"
      >
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Booking
        </button>
      </div>
    </div>
  </div>,
]
`;
